### ISSUE TITLE
The colors looks darker with less details in IINA compared to MPV

### ISSUE THREAD

**Issue by uiryuu** [iina/iina#5124]
### Discussed in https://github.com/iina/iina/discussions/5109

<div type='discussions-op-text'>

<sup>Originally posted by **josepheiba** August  2, 2024</sup>

<img width="1728" alt="Screenshot 2024-08-01 at 11 35 24 PM" src="https://github.com/user-attachments/assets/3b3c82e1-e314-47e5-ac71-26da73aae761">

I have been using MPV, and I always thought it was perfect, but when I tried IINA, the colors didn't seem to match the colors from MPV, IINA is in top left. I can't find anyway in settings to change the quality of the picture to make the colors as vibrant and to make details pop up like MPV.

I tried another player and the colors look similar to MPV.</div>

**Comment by low-batt** [comment]
### Analysis

I am a junior IINA developer operating outside of my area of technical expertice. If you notice something I got wrong, please correct me.

Color is a complex topic. All the little technical details matter. The exact kind of display and it's capabilities. The ambiant light in the room. The viewing angle. How the video was encoded.

I am testing using a 16" MacBookPro18,2 with an Apple M1 Max chip running macOS Ventura 13.6.9. I am running it in closed clamshell attached to a [LG 27" 5K UHD UltraFine™](https://www.lg.com/us/monitors/lg-27md5kl-b-5k-uhd-led-monitor) monitor (DCI-P3 99% Color Gamut / 500nits Brightness). This monitor does not support [Extended Dynamic Range](https://developer.apple.com/documentation/metal/hdr_content/displaying_hdr_content_in_a_metal_layer#3335505).

For audio/video support IINA uses a library from the [mpv project](https://github.com/mpv-player/mpv). It is `libmpv` that generates the video frame to be displayed. IINA is then responsible for displaying the frame. Thus colors displayed by IINA should match mpv.

_However_ mpv is in the middle of major changes, switching from [OpenGL](https://www.opengl.org/) to [Vulkan](https://www.vulkan.org/)/[MoltenVK](https://moltengl.com/moltenvk/)/[Metal](https://developer.apple.com/metal/). The mpv project has not made those changes available to clients using `libmpv`. To compare IINA to mpv the mpv [--vo](https://mpv.io/manual/stable/#video-output-drivers-vo) option must be set to [libmpv](https://mpv.io/manual/stable/#video-output-drivers-libmpv).

Another significant difference between IINA and mpv is that IINA enables hardware decoding by default whereas mpv does not. So we either must disable hardware decoding in IINA or enable it in mpv. As IINA enables it by default I decided to enable it in mpv by setting the [--hwdec](https://mpv.io/manual/stable/#options-hwdec) option to `auto`. The reason why this is important when comparing colors is discussed in the mpv manual in the long entry for  [--hwdec](https://mpv.io/manual/stable/#options-hwdec). Look for the section `Quality reduction with hardware decoding`.

Another difference is the use of ICC profiles. By default IINA uses the ICC profile of the display, mpv does not. IINA's settings on the `Video/Audio` tab I disabled `Load ICC profile` to match up with mpv.
 
IINA 1.3.5 is using mpv 0.35.0 whereas the latest released version of mpv is 0.38.0. The development version of IINA has been upgraded to mpv 0.38.0, so I am using a version of IINA built from the `develop` branch for comparison.
 
In discussion #5109 @josepheiba reported differences with a 10 bit HEVC encoded video. I looked for such a file and ended up downloading the file [Life Untouched HDR10](https://4kmedia.org/life-untouched-hdr-uhd-4k-demo/) from [4K Media](https://4kmedia.org/) to test with.

<details>
<summary><b>mediainfo:</b></summary>

```text
low-batt@gag ~$ mediainfo Life\ Untouched\ 4K\ Demo.mp4 
General
Complete name                            : Life Untouched 4K Demo.mp4
Format                                   : MPEG-4
Format profile                           : Base Media
Codec ID                                 : isom (isom/iso2/mp41)
File size                                : 450 MiB
Duration                                 : 3 min 17 s
Overall bit rate mode                    : Variable
Overall bit rate                         : 19.1 Mb/s
Frame rate                               : 59.940 FPS
Encoded date                             : 2017-06-05 00:13:25 UTC
Tagged date                              : 2017-06-05 00:13:25 UTC
Writing application                      : Lavf57.72.101

Video
ID                                       : 1
Format                                   : HEVC
Format/Info                              : High Efficiency Video Coding
Format profile                           : Main 10@L5.1@Main
Codec ID                                 : hev1
Codec ID/Info                            : High Efficiency Video Coding
Duration                                 : 3 min 17 s
Bit rate                                 : 18.9 Mb/s
Width                                    : 3 840 pixels
Height                                   : 2 160 pixels
Display aspect ratio                     : 16:9
Frame rate mode                          : Variable
Frame rate                               : 59.940 (60000/1001) FPS
Minimum frame rate                       : 59.920 FPS
Maximum frame rate                       : 59.960 FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 10 bits
Bits/(Pixel*Frame)                       : 0.038
Stream size                              : 446 MiB (99%)
Writing library                          : ATEME Titan File 3.7.2 (4.7.2.1001)
Language                                 : English
Encoded date                             : 2017-06-05 00:13:25 UTC
Tagged date                              : 2017-06-05 00:13:25 UTC
Color range                              : Limited
Color primaries                          : BT.2020
Transfer characteristics                 : PQ
Matrix coefficients                      : BT.2020 non-constant
Conformance errors                       : 1
 HEVC                                    : Yes
  General compliance                     : Bitstream parsing ran out of data to read before the end of the syntax was reached, most probably the bitstream is malformed (frame 0, time -00:00:00.017, offset 0x8F4)
Codec configuration box                  : hvcC

Audio
ID                                       : 2
Format                                   : AAC LC
Format/Info                              : Advanced Audio Codec Low Complexity
Codec ID                                 : mp4a-40-2
Duration                                 : 3 min 17 s
Bit rate mode                            : Variable
Bit rate                                 : 127 kb/s
Channel(s)                               : 2 channels
Channel layout                           : L R
Sampling rate                            : 48.0 kHz
Frame rate                               : 46.875 FPS (1024 SPF)
Compression mode                         : Lossy
Stream size                              : 2.96 MiB (1%)
Default                                  : Yes
Alternate group                          : 1
Encoded date                             : 2017-06-05 00:13:25 UTC
Tagged date                              : 2017-06-05 00:13:25 UTC
```
</details>

For testing in IINA's settings I  disabled `Resume last playback position` and enabled `Pause` under `When media is opened` so I could start IINA paused in a specific position. I then started IINA by executing the following command in  [Terminal](https://support.apple.com/guide/terminal/welcome/mac):
```bash
iina-cli ~/Movies/Life\ Untouched\ 4K\ Demo.mp4 -- --start=2
```
For mpv I used this command:
```bash
mpv --no-config --hwdec=auto --pause --start=2 --vo=libmpv  ~/Movies/Life\ Untouched\ 4K\ Demo.mp4 
```

With this video the differences are subtle, but IINA did not match up with mpv.

One difference is that in this case IINA is setting the [CAOpenGLLayer.colorspace](https://developer.apple.com/documentation/quartzcore/caopengllayer/1521873-colorspace) to the color space obtained by calling [CGColorSpaceCreateDeviceRGB](https://developer.apple.com/documentation/coregraphics/1408837-cgcolorspacecreatedevicergb). From the documentation for that method:
> Colors in a device-dependent color space are not transformed or otherwise modified when displayed on an output device—that is, there is no attempt to maintain the visual appearance of a color. As a consequence, colors in a device color space often appear different when displayed on different output devices. For this reason, device color spaces are not recommended when color preservation is important.

Whereas mpv is setting [CAOpenGLLayer.colorspace](https://developer.apple.com/documentation/quartzcore/caopengllayer/1521873-colorspace) to [NSScreen.colorSpace](https://developer.apple.com/documentation/appkit/nsscreen/1388383-colorspace) for the screen the window is on.

Another difference is that IINA is missing support for [cocoa-cb-10bit-context](https://mpv.io/manual/stable/#options-cocoa-cb-10bit-context).

I will be posting a PR that changes IINA to use the screen's color space and adds support for `cocoa-cb-10bit-context`.

This screenshot shows the current development version of IINA before and after the change and mpv.  The difference is subtle. Look closely at the sky in the upper left. Notice the blue is slightly different. Look closely at the orange looking wood in the center.  Notice how it is slightly duller on the left.

| Left | Center | Right |
| :---: | :---: | :---: |
| IINA `develop` | IINA `develop` with Proposed Fix | mpv 0.38.0 |

![comparison](https://github.com/user-attachments/assets/8b49d32a-f1e3-4672-907c-f06899194c6d)

Does this fix the issue @josepheiba is reporting? Not sure. Would need the specific video to tell. What I can say is that these changes will correct some differences related to color.

I glossed over the difference adding `--vo=libmpv` makes. I will come back to that in another post. I need to first get a PR with the changes posted.

**Comment by low-batt** [comment]
On the reason why `--vo=libmpv` must be added when comparing mpv to IINA…

For IINA to be able to use `libmpv` the mpv option [--vo](https://mpv.io/manual/stable/#video-output-drivers-vo) _must_ be set to [libmpv](https://mpv.io/manual/stable/#video-output-drivers-libmpv). The default video driver is `gpu`. As of 0.38.0 the behavior of that driver has changed. It is now using [Vulkan](https://www.vulkan.org/)/[MoltenVK](https://moltengl.com/moltenvk/)/[Metal](https://developer.apple.com/metal/) and [libplacebo](https://github.com/haasn/libplacebo) as seen in this clip from the log file:
```text
[   0.048][v][vo/gpu] Probing for best GPU context.
[   0.048][v][vo/gpu/vulkan] Initializing GPU context 'displayvk'
[   0.048][v][vo/gpu/libplacebo] Initialized libplacebo v7.349.0 (API v349)
```

Many thanks to @krackers for pointing this out.

From the [libplacebo](https://github.com/haasn/libplacebo) README:
> - A colorimetrically accurate **color management** engine with support for soft gamut mapping, ICC profiles, accurate ITU-R BT.1886 emulation, black point compensation, and custom 3DLUTs (.cube).

Switching to `libplacebo` brings in improvements to color management.

The following test invokes mpv 3 times with different options:
```bash
mpv --no-config --hwdec=auto --pause --start=2 --vo=libmpv ~/Movies/Life\ Untouched\ 4K\ Demo.mp4 
```
```bash
mpv --no-config --hwdec=auto --pause --start=2 ~/Movies/Life\ Untouched\ 4K\ Demo.mp4 
```
```bash
mpv --no-config --hwdec=auto --pause --start=2 --vo=gpu-next ~/Movies/Life\ Untouched\ 4K\ Demo.mp4 
```

As can be seen in this screenshot each is different. As in the example above, look at the blue sky in the upper left corner and the orange portion of the tree in the center.

| Left | Center | Right |
| :---: | :---: | :---: |
| mpv `--vo=libmpv` | mpv | mpv `--vo=gpu-next` |

![mpv-comparison](https://github.com/user-attachments/assets/d9ec93a4-71dc-497a-9f27-3f799d575678)

Using mpv [stats](https://mpv.io/manual/stable/#stats) shows differences.

This screenshot shows mpv `--vo=libmpv` is missing some information about the display:
![mpv-libmpv](https://github.com/user-attachments/assets/84c6a82a-bf6d-421c-a732-c392161b1797)

This screenshot shows mpv without options contains more information about the display. Note the format is [rgba16f](https://developer.apple.com/documentation/metal/mtlpixelformat/rgba16float) and the transfer function is [gamma2.2](https://en.wikipedia.org/wiki/Gamma_correction#Computer_displays):

![mpv](https://github.com/user-attachments/assets/1926530a-c131-4968-a1c8-f6206739417a)

This screenshot shows mpv `--vo=gpu-next`. Note the format has changed to [rgb10a2](https://developer.apple.com/documentation/metal/mtlpixelformat/rgb10a2unorm) and the transfer function has changed to [bt.1886](https://en.wikipedia.org/wiki/ITU-R_BT.1886). I believe this is correct for converting HDR to SDR.

![mpv-gpu-next](https://github.com/user-attachments/assets/d7680d91-2183-44aa-9063-355192077778)


The last test I did was to switch to the builtin XDR screen on my MacBook Pro and compare IINA to mpv with EDR enabled.

To get mpv to enable EDR use `--vo=gpu-next` along with [--target-colorspace-hint](https://mpv.io/manual/stable/#options-target-colorspace-hint):
> Automatically configure the output colorspace of the display to pass through the input values of the stream (e.g. for HDR passthrough), if possible. Requires a supporting driver and --vo=gpu-next.

To me IINA and mpv looked the same.

I don't think macOS supports taking a screenshot in a format that supports HDR, such as [JPEG XL](https://jpeg.org/jpegxl/), so the screenshot is badly clipped. You can see that the primaries and transfer of the display have now changed and match the video:

![mpv-hdr](https://github.com/user-attachments/assets/ed7eb306-e534-4ab7-9fe8-24f37955d840)


My conclusion is that this reflects the improvements to color management that mpv is making. Unfortunately IINA can not take advantage of these improvements until they have been made available to clients using `libmpv` and the render API.

**Comment by krackers** [comment]
@low-batt If you use vo=libmpv with mpv you need to set `cocoa-cb-output-csp` to allow the backing CALayer to support EDR I think. If you use gpu or gpu-next which use moltenvk this is done automatically by moltenvk

Also libplacebo is only used for vulkan rendering abstraction with `--vo=gpu`. all rendering/tonemapping/etc. is still handled by the mpv-internal/classic/legacy rendering codepaths.

**Comment by low-batt** [comment]
Great to hear from you. Was hoping you would comment and correct what I got wrong.

Busy day for me. I will have to read over your responses tomorrow. The option `cocoa-cb-output-csp` was added after 0.38.0 was released:
```bash
low-batt@gag ~$ mpv --cocoa-cb-output-csp=auto
Error parsing option cocoa-cb-output-csp (option not found)
Setting commandline option --cocoa-cb-output-csp=auto failed.
Exiting... (Fatal error)
low-batt@gag ~$ 
```

I just tested mpv 0.38.0 with `--vo=gpu-next` and it did not enable EDR on my MacBook Pro with XDR screen. I had to add `--target-colorspace-hint` for EDR to activate.

Gotta run…
